File: src/formatter.js
// mautrix-telegram - A Matrix-Telegram puppeting bridge
// Copyright (C) 2017 Tulir Asokan
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

/**
 * Utility functions to convert between Telegram and Matrix (HTML) formatting.
 * <br><br>
 * <b>WARNING: This module contains headache-causing regular expressions and other duct tape.</b>
 *
 * @module formatter
 */

String.prototype.insert = function(at, str) {
	return this.slice(0, at) + str + this.slice(at)
}

function addSimpleTag(tags, entity, tag, priority = 0) {
	tags.push([entity.offset, `<${tag}>`, -priority])
	tags.push([entity.offset + entity.length, `</${tag}>`, priority])
}

function addTag(tags, entity, tag, attrs, priority = 0) {
	tags.push([entity.offset, `<${tag} ${attrs}>`, -priority])
	tags.push([entity.offset + entity.length, `</${tag}>`, priority])
}

/**
 * Convert a Telegram entity-formatted message to a Matrix HTML-formatted message.
 * <br><br>
 * <b>WARNING: I am not responsible for possible severe headaches caused by reading any part of this function.</b>
 *
 * @param {string} message  The plaintext message.
 * @param {Array}  entities The Telegram formatting entities.
 * @param {MautrixTelegram} app The app main class instance to use when reformatting mentions.
 */
function telegramToMatrix(message, entities, app) {
	const tags = []
	// Decreasing priority counter used to ensure that formattings right next to eachother don't flip like this:
	// *bold*_italic_  -->   <strong>bold<em></strong>italic</em>
	let pc = 9001

	// Convert Telegram formatting entities into a weird custom indexed HTML tag format thingy.
	for (const entity of entities) {
		let url, tag, mxid
		switch (entity._) {
		case "messageEntityBold":
			tag = tag || "strong"
		case "messageEntityItalic":
			tag = tag || "em"
		case "messageEntityCode":
			tag = tag || "code"
			addSimpleTag(tags, entity, tag, --pc)
			break
		case "messageEntityPre":
			pc--
			addSimpleTag(tags, entity, "pre", pc)
			addTag(tags, entity, "code", `class="language-${entity.language}"`, pc + 1)
			break
		case "messageEntityHashtag":
		case "messageEntityBotCommand":
			// TODO bridge bot commands differently?
			addTag(tags, entity, "font", "color=\"blue\"", --pc)
			break
		case "messageEntityMentionName":
			let user = app.matrixUsersByTelegramID.get(entity.user_id)
			if (!user) {
				// TODO this loop step should be made useless
				for (const userByMXID of app.matrixUsersByID.values()) {
					if (userByMXID.telegramUserID === entity.user_id) {
						user = userByMXID
						app.matrixUsersByTelegramID.set(userByMXID.telegramUserID, userByMXID)
						break
					}
				}
			}
			mxid = user ?
				user.userID :
				app.getMXIDForTelegramUser(entity.user_id)
		case "messageEntityMention":
			if (!mxid) {
				const username = message.substr(entity.offset + 1, entity.length - 1)
				for (const userByMXID of app.matrixUsersByID.values()) {
					if (userByMXID._telegramPuppet && userByMXID._telegramPuppet.data.username === username) {
						mxid = userByMXID.userID
						break
					}
				}
				if (!mxid) {
					for (const userByID of app.telegramUsersByID.values()) {
						if (userByID.username === username) {
							mxid = userByID.mxid
							break
						}
					}
				}
			}

			if (!mxid) {
				continue
			}
			addTag(tags, entity, "a", `href="https://matrix.to/#/${mxid}"`)
			break
		case "messageEntityEmail":
			url = url || `mailto:${message.substr(entity.offset, entity.length)}`
		case "messageEntityUrl":
			url = url || message.substr(entity.offset, entity.length)
		case "messageEntityTextUrl":
			url = url || entity.url
			addTag(tags, entity, "a", `href="${url}"`, --pc)
			break
		}
	}

	// Sort tags in a mysterious way (it seems to work, don't touch it!).
	//
	// The important thing is that the tags are sorted last to first,
	// so when replacing by index, the index doesn't need to be adapted.
	tags.sort(([aIndex, , aPriority], [bIndex, , bPriority]) => bIndex - aIndex || aPriority - bPriority)

	// Insert tags into message
	for (const [index, replacement] of tags) {
		message = message.insert(index, replacement)
	}
	return message
}

// Formatting that is converted back to text
const linebreaks = /<br(.*?)>(\n)?/g
const paragraphs = /<p>([^]*?)<\/p>/g
const headers = /<h([0-6])>([^]*?)<\/h[0-6]>/g
const unorderedLists = /<ul>([^]*?)<\/ul>/g
const orderedLists = /<ol>([^]*?)<\/ol>/g
const listEntries = /<li>([^]*?)<\/li>/g

// Formatting that is converted to Telegram entity formatting
const boldText = /<(strong)>()([^]*?)<\/strong>/g
const italicText = /<(em)>()([^]*?)<\/em>/g
const codeblocks = /<(pre><code)>()([^]*?)<\/code><\/pre>/g
const codeblocksWithSyntaxHighlight = /<(pre><code class)="language-(.*?)">([^]*?)<\/code><\/pre>/g
const inlineCode = /<(code)>()(.*?)<\/code>/g
const emailAddresses = /<a href="(mailto):(.*?)">([^]*?)<\/a>/g
const mentions = /<a href="https:\/\/(matrix\.to)\/#\/(@.+?)">(.*?)<\/a>/g
const hyperlinks = /<(a href)="(.*?)">([^]*?)<\/a>/g
const REGEX_CAPTURE_GROUP_COUNT = 3

RegExp.any = function(...regexes) {
	let components = []
	for (const regex of regexes) {
		if (regex instanceof RegExp) {
			components = components.concat(regex._components || regex.source)
		}
	}
	return new RegExp(`(?:${components.join(")|(?:")})`)
}

const regexMonster = RegExp.any(//"g",
		boldText, italicText, codeblocks, codeblocksWithSyntaxHighlight,
		inlineCode, emailAddresses, mentions, hyperlinks)
const NUMBER_OF_REGEXES_EATEN_BY_MONSTER = 8

function regexMonsterMatchParser(match) {
	match.pop() // Remove full string
	const index = match.pop()
	let identifier, arg, text
	for (let i = 0; i < NUMBER_OF_REGEXES_EATEN_BY_MONSTER; i++) {
		if (match[i * REGEX_CAPTURE_GROUP_COUNT]) {
			identifier = match[i * REGEX_CAPTURE_GROUP_COUNT]
			arg = match[(i * REGEX_CAPTURE_GROUP_COUNT) + 1]
			text = match[(i * REGEX_CAPTURE_GROUP_COUNT) + 2]
		}
	}
	return { index, identifier, arg, text }
}

function regexMonsterHandler(identifier, arg, text, index, app) {
	let entity, entityClass, argField
	switch (identifier) {
	case "strong":
		entityClass = "Bold"
		break
	case "em":
		entityClass = "Italic"
		break
	case "pre><code":
	case "pre><code class":
		argField = "language"
		entityClass = "Pre"
		break
	case "code":
		entityClass = "Code"
		break
	case "mailto":
		entityClass = "email"
		// Force text to be the email address
		text = arg
		break
	case "a href":
		if (arg === text) {
			entityClass = "Url"
		} else {
			entityClass = "TextUrl"
			argField = "url"
		}
	case "matrix.to":
		if (app) {
			const match = app.usernameRegex.exec(arg)
			if (!match || match.length < 2) {
				break
			}
			const userID = match[1]

			const user = app.telegramUsersByID.get(+userID)
			if (!user) {
				break
			}

			if (user.username) {
				entityClass = "Mention"
				text = `@${user.username}`
			} else {
				text = user.getDisplayName()
				entity = {
					_: "inputMessageEntityMentionName",
					offset: index,
					length: text.length,
					user_id: {
						_: "inputUser",
						user_id: user.id,
					},
				}
			}
		}
		break
	}
	if (!entity && entityClass) {
		entity = {
			_: `messageEntity${entityClass}`,
			offset: index,
			length: text.length,
		}
		if (argField) {
			entity[argField] = arg
		}
	}
	return { replacement: text, entity }
}

/**
 * Convert a Matrix HTML-formatted message to a Telegram entity-formatted message.
 *
 * @param   {string}                             message The HTML-formatted message.
 * @returns {{message: string, entities: Array}}         The Telegram entity-formatted message.
 */
function matrixToTelegram(message, app) {
	const entities = []

	// First replace all the things that don't get converted into Telegram entities
	message = message.replace(linebreaks, "\n")
	message = message.replace(paragraphs, "$1\n")
	message = message.replace(headers, (_, count, text) => `${"#".repeat(count)} ${text}`)
	message = message.replace(unorderedLists, (_, list) => list.replace(listEntries, "- $1"))
	message = message.replace(orderedLists, (_, list) => {
		let n = 0
		return list.replace(listEntries, (fullMatch, text) => `${++n}. ${text}`)
	})

	const regexMonsterReplacer = (match, ...args) => {
		const { index, identifier, arg, text } = regexMonsterMatchParser(args)
		if (!identifier) {
			// This shouldn't happen
			console.warn(`Warning: Match found but parsing failed for match "${match}"`)
			return match
		}
		const { replacement, entity } = regexMonsterHandler(identifier, arg, text, index, app)
		if (entity) {
			entities.push(entity)
		}
		return replacement || text
	}

	// We replace matches iteratively to make sure the indexes of matches are correct.
	let oldMessage = message
	message = message.replace(regexMonster, regexMonsterReplacer)
	while (oldMessage !== message) {
		oldMessage = message
		message = message.replace(regexMonster, regexMonsterReplacer)
	}

	return { message, entities }
}

module.exports = { telegramToMatrix, matrixToTelegram }

File: mautrix_telegram/formatter.py
# -*- coding: future_fstrings -*-
# mautrix-telegram - A Matrix-Telegram puppeting bridge
# Copyright (C) 2018 Tulir Asokan
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.
from html import escape, unescape
from html.parser import HTMLParser
from collections import deque
import re
import logging

from matrix_client.errors import MatrixRequestError

from telethon.tl.types import *

from . import user as u, puppet as p
from .db import Message as DBMessage

log = logging.getLogger("mau.formatter")


# region Matrix to Telegram

class MessageEntityReply(MessageEntityUnknown):
    def __init__(self, offset=0, length=0, msg_id=0):
        super().__init__(offset, length)
        self.msg_id = msg_id


class MatrixParser(HTMLParser):
    mention_regex = re.compile("https://matrix.to/#/(@.+)")
    reply_regex = re.compile(r"https://matrix.to/#/(!.+?)/(\$.+)")

    def __init__(self, tg_space=None):
        super().__init__()
        self._tg_space = tg_space
        self.text = ""
        self.entities = []
        self._building_entities = {}
        self._list_counter = 0
        self._open_tags = deque()
        self._open_tags_meta = deque()
        self._previous_ended_line = True
        self._building_reply = False

    def handle_starttag(self, tag, attrs):
        self._open_tags.appendleft(tag)
        self._open_tags_meta.appendleft(0)
        attrs = dict(attrs)
        entity_type = None
        args = {}
        if tag == "strong" or tag == "b":
            entity_type = MessageEntityBold
        elif tag == "em" or tag == "i":
            entity_type = MessageEntityItalic
        elif tag == "code":
            try:
                pre = self._building_entities["pre"]
                try:
                    pre.language = attrs["class"][len("language-"):]
                except KeyError:
                    pass
            except KeyError:
                entity_type = MessageEntityCode
        elif tag == "pre":
            entity_type = MessageEntityPre
            args["language"] = ""
        elif tag == "a":
            try:
                url = attrs["href"]
            except KeyError:
                return
            mention = self.mention_regex.search(url)
            reply = self.reply_regex.search(url)
            if mention:
                mxid = mention.group(1)
                user = p.Puppet.get_by_mxid(mxid, create=False)
                if not user:
                    user = u.User.get_by_mxid(mxid, create=False)
                    if not user:
                        return
                if user.username:
                    entity_type = MessageEntityMention
                    url = f"@{user.username}"
                else:
                    entity_type = MessageEntityMentionName
                    args["user_id"] = user.tgid
            elif reply and self._tg_space and (len(self.entities) == 0
                                               and len(self._building_entities) == 0):
                room_id = reply.group(1)
                message_id = reply.group(2)
                message = DBMessage.query.filter(DBMessage.mxid == message_id,
                                                 DBMessage.mx_room == room_id,
                                                 DBMessage.tg_space == self._tg_space
                                                 ).one_or_none()
                if not message:
                    return
                entity_type = MessageEntityReply
                args["msg_id"] = message.tgid
                self._building_reply = True
                url = None
            elif url.startswith("mailto:"):
                url = url[len("mailto:"):]
                entity_type = MessageEntityEmail
            else:
                if self.get_starttag_text() == url:
                    entity_type = MessageEntityUrl
                else:
                    entity_type = MessageEntityTextUrl
                    args["url"] = url
                    url = None
            self._open_tags_meta.popleft()
            self._open_tags_meta.appendleft(url)

        if entity_type and tag not in self._building_entities:
            self._building_entities[tag] = entity_type(offset=len(self.text), length=0, **args)

    def _list_depth(self):
        depth = 0
        for tag in self._open_tags:
            if tag == "ol" or tag == "ul":
                depth += 1
        return depth

    def handle_data(self, text):
        text = unescape(text)
        if self._building_reply:
            return
        previous_tag = self._open_tags[0] if len(self._open_tags) > 0 else ""
        list_format_offset = 0
        if previous_tag == "a":
            url = self._open_tags_meta[0]
            if url:
                text = url
        elif len(self._open_tags) > 1 and self._previous_ended_line and previous_tag == "li":
            list_type = self._open_tags[1]
            indent = (self._list_depth() - 1) * 4 * " "
            text = text.strip("\n")
            if len(text) == 0:
                return
            elif list_type == "ul":
                text = f"{indent}* {text}"
                list_format_offset = len(indent) + 2
            elif list_type == "ol":
                n = self._open_tags_meta[1]
                n += 1
                self._open_tags_meta[1] = n
                text = f"{indent}{n}. {text}"
                list_format_offset = len(indent) + 3
        for tag, entity in self._building_entities.items():
            entity.length += len(text.strip("\n"))
            entity.offset += list_format_offset

        if text.endswith("\n"):
            self._previous_ended_line = True
        else:
            self._previous_ended_line = False

        self.text += text

    def handle_endtag(self, tag):
        try:
            self._open_tags.popleft()
            self._open_tags_meta.popleft()
        except IndexError:
            pass
        if tag == "a":
            self._building_reply = False
        if (tag == "ul" or tag == "ol") and self.text.endswith("\n"):
            self.text = self.text[:-1]
        entity = self._building_entities.pop(tag, None)
        if entity:
            self.entities.append(entity)


def matrix_to_telegram(html, tg_space=None):
    try:
        parser = MatrixParser(tg_space)
        parser.feed(html)
        return parser.text, parser.entities
    except Exception:
        log.exception("Failed to convert Matrix format:\nhtml=%s", html)


# endregion
# region Telegram to Matrix

def telegram_event_to_matrix(evt, source, native_replies=False, main_intent=None):
    text = evt.message
    html = telegram_to_matrix(evt.message, evt.entities) if evt.entities else None

    if evt.fwd_from:
        if not html:
            html = escape(text)
        from_id = evt.fwd_from.from_id
        user = u.User.get_by_tgid(from_id)
        if user:
            fwd_from = f"<a href='https://matrix.to/#/{user.mxid}'>{user.mxid}</a>"
        else:
            puppet = p.Puppet.get(from_id, create=False)
            if puppet and puppet.displayname:
                fwd_from = f"<a href='https://matrix.to/#/{puppet.mxid}'>{puppet.displayname}</a>"
            else:
                user = source.client.get_entity(from_id)
                if user:
                    fwd_from = p.Puppet.get_displayname(user, format=False)
                else:
                    fwd_from = None
        if not fwd_from:
            fwd_from = "Unknown user"
        html = (f"Forwarded message from <b>{fwd_from}</b><br/>"
                + f"<blockquote>{html}</blockquote>")

    if evt.reply_to_msg_id:
        space = evt.to_id.channel_id if isinstance(evt, Message) and isinstance(evt.to_id,
                                                                                PeerChannel) else source.tgid
        msg = DBMessage.query.get((evt.reply_to_msg_id, space))
        if msg:
            if native_replies:
                quote = f"<a href=\"https://matrix.to/#/{msg.mx_room}/{msg.mxid}\">Quote<br></a>"
            else:
                try:
                    event = main_intent.get_event(msg.mx_room, msg.mxid)
                    content = event["content"]
                    body = content["formatted_body"] if "formatted_body" in content else content["body"]
                    reply_to = f"<a href='https://matrix.to/#/{event['sender']}'>event['sender']</a>"
                    quote = f"Reply to {reply_to}<blockquote>{body}</blockquote>"
                except (ValueError, KeyError, MatrixRequestError):
                    quote = "Reply to someone (failed to fetch message)<br/>"
            if html:
                html = quote + html
            else:
                html = quote + escape(text)

    if html:
        html = html.replace("\n", "<br/>")

    return text, html


def telegram_to_matrix(text, entities):
    try:
        return _telegram_to_matrix(text, entities)
    except Exception:
        log.exception("Failed to convert Telegram format:\n"
                      "message=%s\n"
                      "entities=%s",
                      text, entities)


def _telegram_to_matrix(text, entities):
    if not entities:
        return text
    html = []
    last_offset = 0
    for entity in entities:
        if entity.offset > last_offset:
            html.append(escape(text[last_offset:entity.offset]))
        elif entity.offset < last_offset:
            continue

        skip_entity = False
        entity_text = escape(text[entity.offset:entity.offset + entity.length])
        entity_type = type(entity)

        if entity_type == MessageEntityBold:
            html.append(f"<strong>{entity_text}</strong>")
        elif entity_type == MessageEntityItalic:
            html.append(f"<em>{entity_text}</em>")
        elif entity_type == MessageEntityCode:
            html.append(f"<code>{entity_text}</code>")
        elif entity_type == MessageEntityPre:
            if entity.language:
                html.append("<pre>"
                            + f"<code class='language-{entity.language}'>{entity_text}</code>"
                            + "</pre>")
            else:
                html.append(f"<pre><code>{entity_text}</code></pre>")
        elif entity_type == MessageEntityMention:
            username = entity_text[1:]

            user = u.User.find_by_username(username)
            if user:
                mxid = user.mxid
            else:
                puppet = p.Puppet.find_by_username(username)
                mxid = puppet.mxid if puppet else None
            if mxid:
                html.append(f"<a href='https://matrix.to/#/{mxid}'>{entity_text}</a>")
            else:
                skip_entity = True
        elif entity_type == MessageEntityMentionName:
            user = u.User.get_by_tgid(entity.user_id)
            if user:
                mxid = user.mxid
            else:
                puppet = p.Puppet.get(entity.user_id, create=False)
                mxid = puppet.mxid if puppet else None
            if mxid:
                html.append(f"<a href='https://matrix.to/#/{mxid}'>{entity_text}</a>")
            else:
                skip_entity = True
        elif entity_type == MessageEntityEmail:
            html.append(f"<a href='mailto:{entity_text}'>{entity_text}</a>")
        elif entity_type == MessageEntityUrl:
            html.append(f"<a href='{entity_text}'>{entity_text}</a>")
        elif entity_type == MessageEntityTextUrl:
            html.append(f"<a href='{escape(entity.url)}'>{entity_text}</a>")
        elif entity_type == MessageEntityBotCommand:
            html.append(f"<font color='blue'>!{entity_text[1:]}")
        elif entity_type == MessageEntityHashtag:
            html.append(f"<font color='blue'>{entity_text}</font>")
        else:
            skip_entity = True
        last_offset = entity.offset + (0 if skip_entity else entity.length)
    html.append(text[last_offset:])

    return "".join(html)

# endregion
